### DISCUSSION TITLE
Introduce @Nullable @Notnull annotations.

### DISCUSSION THREAD

**Discussion by mattisonchao** [apache/pulsar#18028] [Ideas]
Currently, we have many util methods and public methods.
For project quality, I would like to introduce `@Notnull` and `@Nullable` annotations to help contributors easily understand methods and avoid NPE.

I expect we introduce two annotations for Pulsar.

```java
@Documented
@TypeQualifierDefault({ ElementType.PARAMETER, ElementType.FIELD, ElementType.METHOD, ElementType.TYPE_USE })
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.PACKAGE)
@javax.annotation.Nonnull
@org.jetbrains.annotations.NotNull
public @interface NotNull {
}
```

```java
@Documented
@TypeQualifierDefault({ ElementType.PARAMETER, ElementType.FIELD, ElementType.METHOD, ElementType.TYPE_USE })
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.PACKAGE)
@javax.annotation.Nullable
@org.jetbrains.annotations.Nullable
public @interface Nullable {
}
```

NPE problems: https://github.com/apache/pulsar/pulls?q=+NPE+

**Comment by BewareMyPower** [comment]
Cannot agree more.

**Comment by nahguam** [comment]
I'm in favour of greater clarity but why create new annotations and not just use the javax or jetbrains equivalents directly? Or indeed the lombok equivalents.

**Comment by Jason918** [comment]
Seems we are combining "@javax.annotation.x" and "@org.jetbrains.annotations.x"? Can you bring more context on these two?
